### PR TITLE
[20_13] Do not add duplicated values to copy_always to avoid stack overflow

### PR DESCRIPTION
**Pull request by wind1900**
## What
Add check before adding new values to `edit_interface_rep::copy_always` in function `edit_interface_rep::apply_changes()`.
New rectangle will be added to `copy_always` if `copy_always` is empty/nil or the first item of `copy_always` is not equal to newly added value.
## Why
Since `apply_changes()` will be called when the window receives any event, `copy_always` will become very large and cause stack overflow when it's cleared.
The `copy_always` is a linked list and when it's cleared, the destructor is called from head to tail recursively, causing stack overflow.
## How to test your changes?
That's the potential risk: I can only test it on Windows with few scenarios, I don't know the result in Linux, Mac.

already run `xmake run --yes -vD --group=tests` and `xmake run --yes -vD --group=integration_tests`

The investigations below are all in Windows environment.
### Where is the code to clear `copy_always`?
In `edit_interface_rep::scroll_to()` and `edit_interface_rep::set_extents()`, by using `copy_always = rectangles ()`.

### When will `copy_always` be cleared?
`scroll_to()` is called when the cursor is moved out of current view.
I can't add breakpoint in `set_extents()`, I don't know when will it will be called.

### What's the purpose of `copy_always`?
The comment says it's used for wiping out cursor.
However I think it doesn't do anything now.
I comment out the code that uses it, get nothing wrong.
But removing it is more dangerous.

### Where is `copy_always` used?
In `edit_interface_rep::draw_pre()` and `edit_interface_rep::draw_with_shadow()`.

In `draw_pre()`, it will go through all rectangles in `copy_always` and call `win->put_shadow()`, but it doesn't draw anything at all.
Because in `qt_renderer_rep::put_shadow()`, `painter == static_cast<qt_renderer_rep*> (ren)->painter` is true and returns directly.

In `draw_with_shadow()`, it's only used if `gui_interrupted()` returns true, but `gui_interrupted()` doesn't return true in my machine.

These shadow functions are related to the double buffering, the behavior may be different in other OS.

In summary, I don't think it's necessary to store all the old rectangles in `copy_always`.
The latest ones `ocr`(old cursor) and `ncr`(new cursor) should be enough for cursor drawing stuff.

### CODE REVIEW

File: src/Edit/Interface/edit_interface.cpp [hunk 1]
```diff
@@ -828,6 +828,7 @@ edit_interface_rep::apply_changes () {
     if (tremble_count > 3) dw= (1 + min (tremble_count - 3, 25)) * 2 * pixel;
     SI /*P1= zpixel,*/ P2= 2 * zpixel, P3= 3 * zpixel;
     cursor             cu= get_cursor ();
+    copy_always          = rectangles ();
```
[the remark this answers is not in the document]
Comment (da-liii) [reply src/Edit/Interface/edit_interface.cpp:831]: 上面第一个rectangle是old cursor，第二个rectangle是new cursor

**Comment by da-liii** [comment]
CI已经修复，合并branch-1.2，再push一下，这个pr就可以正常通过CI了

File: src/Edit/Interface/edit_interface.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (da-liii) [reply src/Edit/Interface/edit_interface.cpp:831]: 我建议下面的逻辑改成：
1. 计算ocr
2. 计算ncr
3. 如果ocr和ncr相等，那么invalidate只调用一次（这里假设invalidate的入参一样，调用两次效果是一样的）
4. 如果ocr和ncr相等且和copy_always的栈顶相等，那么不需要更新copy_always，否则按照原来的逻辑，在copy_always这个栈中入栈两个元素

[the remark this answers is not in the document]
Comment (da-liii) [reply src/Edit/Interface/edit_interface.cpp:831]: 第四点是关键，第三点是否要改为调用一次invalidate，具体看invalidate的效果决定

[the remark this answers is not in the document]
Comment (wind1900) [reply src/Edit/Interface/edit_interface.cpp]: invalidate我保持原样不动它了, 现在改成4了, 栈顶相等就不更新

**Review by da-liii — APPROVED** [review]
LGTM
